core/src/main/java/com/linecorp/armeria/client/endpoint/WeightBasedRandomEndpointSelector.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+import javax.annotation.Nullable;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableList.Builder;
+
+import com.linecorp.armeria.client.Endpoint;
+
+final class WeightBasedRandomEndpointSelector {
+    private static final AtomicIntegerFieldUpdater<WeightBasedRandomEndpointSelector> removingEntryUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(WeightBasedRandomEndpointSelector.class, "removingEntry");
+
+    private final List<Entry> entries;
+    private final long totalWeight;
+    private final List<Entry> currentEntries;
+    private long currentTotalWeight;
+
+    private volatile int removingEntry;
+
+    WeightBasedRandomEndpointSelector(List<Endpoint> endpoints) {
+        final Builder<Entry> builder = ImmutableList.builder();
+
+        long totalWeight = 0;
+        for (Endpoint endpoint : endpoints) {
+            builder.add(new Entry(endpoint, totalWeight));
+            totalWeight += endpoint.weight();
+        }
+        this.totalWeight = totalWeight;
+        currentTotalWeight = totalWeight;
+        entries = builder.build();
+        currentEntries = new CopyOnWriteArrayList<>(entries);
+    }
+
+    @VisibleForTesting
+    List<Entry> entries() {
+        return entries;
+    }
+
+    @Nullable
+    Endpoint selectEndpoint() {
+        if (entries.isEmpty()) {
+            return null;
+        }
+        for (;;) {
```

Comment:
It's unclear from the code that this can't become an infinite loop. Is there a way to restructure this to not have a loop? Otherwise probably should add comments here on what the loop is for.

Reply:
Let me add a comment for this. 😉